nemo/backends/pytorch/actions.py

```diff
@@ -87,137 +89,72 @@ def __init__(
-            if nmtensor.producer_args is not None and nmtensor.producer_args != {}:
-                for _, new_nmtensor in nmtensor.producer_args.items():
-                    if new_nmtensor not in processed_nmtensors:
-                        # put in the start of list
-                        hooks_lst.insert(0, new_nmtensor)
-
-        all_node_with_output = []
-        # Iterate over all_nodes to create new nodes that include its output
-        # now all nodes have (module, input tensors, output tensors)
-        for node in all_nodes:
-            all_node_with_output.append(tuple((node[0], node[1], all_nodes[node])))
-
-        processed_nodes = []
-        while len(all_node_with_output) > 0:
-            for node in all_node_with_output.copy():
-                # if node's in_degree is zero it can be added to
-                # _top_sorted_modules
-                # this will also reduce in_degree of its children
-                if is_in_degree_zero(node, processed_nodes):
-                    _top_sorted_modules.append(node)
-                    processed_nodes.append((node[0], node[1]))
-                    all_node_with_output.remove(node)
-
-        # Create top_sorted_modules aka callchain
-        top_sorted_modules = []
-        for i, m in enumerate(_top_sorted_modules):
-            top_sorted_modules.append((m[0], dict(m[1]), m[2]))
-            # Ensure that there is only one dataset in callchain
-            if i > 0 and isinstance(m[0], DataLayerNM):
-                raise ValueError("There were more than one DataLayer NeuralModule inside your DAG.")
+    @epoch.setter
+    def epoch(self, epoch):
+        self._epoch = epoch
+
+    @property
+    @deprecated(version="0.12", explanation="epoch_num has been deprecated in favour of epoch.")
+    def epoch_num(self):
+        return self._epoch
+
+    @epoch_num.setter
+    @deprecated(version="0.12", explanation="epoch_num has been deprecated in favour of epoch.")
+    def epoch_num(self):
+        return self._epoch
+
+    def __get_top_sorted_modules_and_dataloader(self, hook: List[NmTensor]):
+        """A function that accepts a list of NmTensors that need to be computed and constructs a call DAG that starts
+        from a datalayerNM and can be used to compute the NmTensors.
+
+        args:
+            leaf_nmtensors (List[NmTensors]): The tensors to be computed
+
+        returns:
+            top_sorted_modules: the callchain DAG
+            tdataset: the datalayer at the top of the callchain
+        """
+        top_sorted_modules = topological_sort_from_leaves(hook)
 
         if not isinstance(top_sorted_modules[0][0], DataLayerNM):
             raise ValueError("The first module in your DAG was not a DataLayer NeuralModule.")
 
         tdataset = top_sorted_modules[0][0].dataset
 
-        # populate self.module_reference_table
         for m in top_sorted_modules:
             if m[0].factory is None and self._local_rank is not None:
```

Comment:
Minor comment: I think that the first part of this condition simply cannot be True now, ever - factory is always set now in NM __init__()